Glm docs (#2504)
nilearn/nilearn: doc/glm/second_level_model.rst

.. _second_level_model:

===================
Second level models
===================

.. topic:: **Page summary**

>    Second level models are used to take fMRI analysis to the group level. Once individual subject data has been processed in a common space (e.g. MNI, Talairach, or subject average), the data can be grouped and statistical tests  performed to make broader inferences on fMRI activity.

Reviewer: We should say that this page focuses on running a second-level model with nilearn (and make sure that the page is indeed framed this way).

One-sample testing: unpaired
----------------------------

In one-sample testing, activity associated with a predictor variable is tested against some form of baseline activity. Therefore, predictors must be defined in order to set up a one-sample t-test. For example, consider the first level contrast of left-vs-right button press, which can be downloaded using :func:`nilearn.datasets.fetch_localizer_contrasts`::

  from nilearn.datasets import fetch_localizer_contrasts
  n_subjects = 16
  data = fetch_localizer_contrasts(["left vs right button press"], n_subjects,
                               get_tmaps=True)

A first level model has been fit to this data and the contrasts maps for left-vs-right button press activation are available in the `cmaps` parameter. The second level model is fit using these contrast maps thus::

  from nilearn.stats.second_level_model import make_second_level_design_matrix
  second_level_input = data['cmaps']

>   design_matrix = pd.DataFrame([1] * len(second_level_input), columns=['intercept'])

Reviewer: Here also, there shouldn't be code snippets. Just pointers to functions, classes, and examples.

  from nilearn.datasets import fetch_localizer_contrasts
  n_subject = 16
  sample_vertical = fetch_localizer_contrasts(
      ["vertical checkerboard"], n_subjects, get_tmaps=True)
  sample_horizontal = fetch_localizer_contrasts(
      ["horizontal checkerboard"], n_subjects, get_tmaps=True)
  second_level_input = sample_vertical['cmaps'] + sample_horizontal['cmaps']

Set up the design matrix::

  import numpy as np
  condition_effect = np.hstack(([1] * n_subjects, [- 1] * n_subjects))
  subject_effect = np.vstack((np.eye(n_subjects), np.eye(n_subjects)))
  subjects = ['S%02d' % i for i in range(1, n_subjects + 1)]
  design_matrix = pd.DataFrame(
      np.hstack((condition_effect[:, np.newaxis], subject_effect)),
      columns=['vertical vs horizontal'] + subjects)

As with first level models, the design matrix can be visualized using :func:`nilearn.reporting.plot_design_matrix()`

.. image:: ../auto_examples/05_glm_second_level_models/images/sphx_glr_plot_second_level_two_sample_test_001.png
   :target: ../auto_examples/05_glm_second_level_models/plot_second_level_two_sample_test.html

> 

Reviewer 2: Actually this one is a paired t-test. Should be decribed as such.

Author: I agree. The label and description for this example should be changed in the Examples section as well.

Thresholding statistical maps
=============================

Nilearn's statistical plotting functions provide simple thresholding functionality. For instance, functions like :func:`nilearn.plotting.plot_glass_brain` or :func:`nilearn.plotting.plot_glass_brain` have an argument called `threshold` that only show voxels with a value that is over the threshold provided. Thresholding examples are available here: :ref:`sphx_glr_auto_examples_05_glm_second_level_models_plot_second_level_one_sample_test.py` and :ref:`sphx_glr_auto_examples_05_glm_second_level_models_plot_thresholding.py`.

Multiple comparisons correction
===============================

As discussed in the :ref:`Multiple comparisons` section, the issue of multiple comparisons is important to address with statistical analysis of fMRI data. Nilearn provides parametric and non-parametric tools to address this issue.

Refer to the example :ref:`sphx_glr_auto_examples_05_glm_second_level_models_plot_thresholding.py` for a guide to applying FPR, FDR and FWER corrections. These corrections are applied using the :func:`nilearn.stats.map_threshold` function.

> Within an activated cluster, not all voxels reprepsent true activation. To estimate true positives within a cluster, Nilearn provides the :func:`nilearn.stats.cluster_level_inference` function. An example with usage information is available here: :ref:`sphx_glr_auto_examples_05_glm_second_level_models_plot_proportion_activated_voxels.py`

Reviewer 2: ```suggestion
Within an activated cluster, not all voxels represent true activation. To estimate true positives within a cluster, Nilearn provides the :func:`nilearn.stats.cluster_level_inference` function. An example with usage information is available here: :ref:`sphx_glr_auto_examples_05_glm_second_level_models_plot_proportion_activated_voxels.py`
```

  subjects = ['S%02d' % i for i in range(1, n_subjects + 1)]
  design_matrix = pd.DataFrame(
      np.hstack((condition_effect[:, np.newaxis], subject_effect)),
      columns=['vertical vs horizontal'] + subjects)

As with first level models, the design matrix can be visualized using
:func:`nilearn.reporting.plot_design_matrix()`

.. image:: ../auto_examples/05_glm_second_level_models/images/sphx_glr_plot_second_level_two_sample_test_001.png
   :target: ../auto_examples/05_glm_second_level_models/plot_second_level_two_sample_test.html#sphx-glr-auto-examples-05-glm-second-level-models-plot-second-level-two-sample-test-py

Fitting a second level model
============================

Once the design matrix has been set up, fitting the model is simple. The class that
provides tools for this is :class:`nilearn.stats.second_level_model.SecondLevelModel`.
The :func:`nilearn.stats.second_level_model.SecondLevelModel.fit` function fits the
model::

  from nilearn.stats.second_level_model import SecondLevelModel
  second_level_model = SecondLevelModel().fit(
      second_level_input, design_matrix=design_matrix)

While contrast maps can be computed using the

> :func:`nilearn.stats.second_level_model.compute_contrast` function. For instance, the

Reviewer 3: The current CI error is because this target isn't defined. Looking [at the current master branch](https://github.com/nilearn/nilearn/blob/b5966b944f33e1abff9ab46e1ca450f35e95ed55/nilearn/stats/__init__.py#L4-L5), I think what you want here is:

```suggestion
:func:`nilearn.stats.contrasts.compute_contrast` function. For instance, the
```

.. _second_level_model:

===================
Second level models
===================

.. topic:: **Page summary**

   Second level models in Nilearn are used to perform group-level analyses on fMRI data. Once individual

>    subjects has been processed in a common space (e.g. MNI, Talairach, or subject average), the data can

Reviewer 2: ```suggestion
   subjects have been processed in a common space (e.g. MNI, Talairach, or subject average), the data can
```
